### PR TITLE
Data Components fixes/changes

### PR DESCRIPTION
**Pull request by dlasky**
Changes resulting from dogfooding + some fixes to existing architecture that hadn't been properly tested.

@shuwen @anthonykoerber @jcmoore

### CODE REVIEW

**Comment by dlasky** [comment]
There is a regression that affects data binding for model/collection caused by event namespacing, will reopen when fixed.

**Comment by dlasky** [comment]
the event stuff was actually caused by the namespacing breaking the binding by modifying the event type passed from binding fire calls

File: src/shared/js/ajax.html [hunk 1]
```diff
@@ -247,13 +247,18 @@
 			return output;
 		},
 
-		fire: function(type, detail) {
-			if (this._eventPrefix) {
+		fire: function(type, detail, options) {
+			if (this._eventPrefix && DataUtils.getPathValue('prefix', options)) {
```
Comment (anthonykoerber) [inline src/shared/js/ajax.html]: Looks like this needs to be `StrandLib.DataUtils.getPathValue('prefix', options)`

Comment (dlasky) [inline src/shared/js/ajax.html]: good catch but those refs should be extracted to the IIFE so it reads cleaner. see [commit](https://github.com/MediaMath/strand/commit/d6666791a8160202faecafc601c37f6929971ef5)

**Comment by dlasky** [comment]
cleaned up the ajax example to make output prettier, also determined what was causing the binds to fail. The behavior export was happening after the event binding was being applied causing some of the event stuff to [break](https://github.com/MediaMath/strand/commit/ab458e7a95ebd21b84a0f890f88770f243470e43).

File: src/shared/behaviors/collection.html [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+		add: function(model, silent, force) {
+			this._silent = silent;
+
+			if (!model.cId || model.cId === -1) {
+				_generateCid(this, model);
+			}
+			var check = this.data.filter(function(m) {
+				if (m.cId && model.cId) { return m.cId === model.cId; }
+				if (m.mId && model.mId) { return m.mId === model.mId; }
```
Comment (jcmoore) [inline src/shared/behaviors/collection.html]: Is the above intentionally preferred over something like:

```
if (m.cId && model.cId && m.cId === model.cId) { return true; }
if (m.mId && model.mId && m.mId === model.mId) { return true; }
```

In other words(, and although it may be a somewhat exceptional edge-case), the question is "provided `.cId` exists, is a `.cId` mismatch strictly more important than a `.mId`match or not?"

Comment (dlasky) [inline src/shared/behaviors/collection.html]: cid = collection id which is assigned locally, so a mismatch is catastrophic in many ways.

File: src/shared/behaviors/model.html [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+		},
+
+		toJSON: function() {
+			return JSON.stringify(this.data);
+		},
+
+		// _handleResponse: function(e) {
+		// 	this.data = DataUtils.getPathValue("detail.marshalled",e) || DataUtils.getPathValue("detail.response",e);
+		// 	this.fire("changed", {instance:this, data:this.data});
+		// },
```
Comment (anthonykoerber) [inline src/shared/behaviors/model.html]: Do we want to keep all of the commented bits?

File: src/shared/behaviors/domsyncable.html [hunk 4]
```diff
@@ -14,12 +14,12 @@
...
-			params: obj.queryparam || [],
+			query: obj.queryparam || [],
 			urlParams: obj.urlparam || [],
 			headers: obj.header || []
 		};
 	}
-	
+
 	var Sync = StrandLib.Sync;
 	var Ajax = StrandLib.Ajax;
```
Comment (anthonykoerber) [inline src/shared/behaviors/domsyncable.html]: Do we need Sync & Ajax here?

**Comment by anthonykoerber** [comment]
Other than the few minor comments, and the other concerns which were discussed and resolved already - looking good to me. 

TL;DR +1

File: src/shared/js/request.html [hunk 5]
```diff
@@ -106,21 +110,21 @@
...
+				this.fire('preflight',detail);
 			}
 		},
 
 		handleProgress: function(e) {
 			var detail = {
-					percent: e.totalSize/e.position,
-					total: e.totalSize,
-					current: e.position
+					percent: e.total || e.totalSize/e.loaded || e.position,
```
Comment (jcmoore) [inline src/shared/js/request.html]: Watch out order of operations here (parens around the `||`ed or subexpressions) and make sure you want `total/loaded` and not `loaded/total`

**Comment by dlasky** [comment]
@jcmoore see [commit](https://github.com/MediaMath/strand/commit/e3a71fc4e6037fb5037d59bebf0d7c5165be68b7)

File: src/shared/js/request.html (hunk 5 above)
Comment (jcmoore) [inline src/shared/js/request.html]: Looks good to me, +1

**Comment by dlasky** [comment]
rebased to develop
